stream_alert/classifier/parsers.py

```diff
@@ -0,0 +1,808 @@
+            return
+
+        LOGGER.debug('Extracting envelope keys')
+        return {
+            key: payload[key]
+            for key in self.envelope_schema
+            if key in payload  # This is fine since some of these may be optional
+        }
+
+    def _json_path_records(self, payload):
+        """Extract records from the original json payload using a provided JSON path
+
+        Args:
+            payload (dict): The parsed json data
+
+        Returns:
+            list: A list of JSON records extracted via JSON path
+        """
+        # Handle jsonpath extraction of records
+        LOGGER.debug('Parsing records with JSONPath: %s', self.json_path)
+
+        return jmespath.search(self.json_path, payload)
+
+    def parse(self, data):
+        """Main parser method to be handle parsing of the passed data
+
+        Args:
+            data (str|dict): Data to be parsed.
+
+        Returns:
+            bool: True if any records were parsed using this schema, False otherwise
+        """
+        # Ensure the schema is defined properly. Invalid schemas will not be used
+        if not self._validate_schema():
+            LOGGER.error(
+                'Schema definition is not valid (%s):\n%s', self._schema_type, self.schema)
+            return False
+
+        data_copy = None
+        # If the data is a mutable object, copy it since
+        # the parse(s) can highly mutate the input
+        if isinstance(data, dict):
+            data_copy = deepcopy(data)
+
+        # If the data is a string, attempt to load it to json,
+        # falling back on a string type on error
+        elif isinstance(data, basestring):
+            try:
+                data_copy = json.loads(data)
+            except (ValueError, TypeError) as err:
+                LOGGER.debug('Data is not valid json: %s', err.message)
+                data_copy = data
+
+        # Check to make sure any non-optional envelope keys exist before proceeding
+        if not self._key_check(data_copy, self.envelope_schema, self.optional_envelope_keys, True):
+            return False
+
+        # Get the envelope and try to convert the value to the proper type(s)
+        envelope = self._extract_envelope(data_copy)
+        if not self._convert_type(envelope, self.envelope_schema, self.optional_envelope_keys):
+            return False
+
+        # Add the optional envelope keys to record once at the beginning
+        self._add_optional_keys(envelope, self.envelope_schema, self.optional_envelope_keys)
+
+        for record, valid in self._parse(data_copy):
+            valid = valid and self._key_check(record, self.schema, self.optional_top_level_keys)
+            valid = valid and self._convert_type(record, self.schema, self.optional_top_level_keys)
+            valid = valid and self._matches_log_patterns(record, self.log_patterns)
+            self._add_parse_result(record, valid, envelope)
+
+        return self.valid
```

Comment:
where is this set?

Reply:
this is a computed property (not set explicitly) and is considered true if this schema resulted in any parsed records. see:

https://github.com/airbnb/streamalert/pull/821/files#diff-39b8f0570d770502a95e06c76883fceeR121